KEP-1880 Multiple ServiceCIDR ( and IPAddress allocation)

aojea opened the pull request:
/kind feature
/kind api-change


[After adding the new IPAddress object,](https://github.com/kubernetes/kubernetes/pull/115075) this PR adds a new ServiceCIDR object that allows to dynamically configure the ranges used by Services in a cluster.

```release-note
Add a new ServiceCIDR type that allows to dynamically configure the cluster range used to allocate Service ClusterIPs addresses
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:


```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs
```

#### Information for reviewers

The PR is split in independent commits

- API definition of ServiceCIDR (it uses a Ready condition)
- Integration API in the apiserver
- apiserver controller to handle the default serviceCIDR
- introduction of an IP radix tree structure that is the base of the functionality of all the controllers
- extend the existing IPAllocator (allocator that uses the IPAddress object) to handle multiple and dynamic ServiceCIDRs
- extend the existing repair loop for the IPAllocator to handle multiple ServiceCIDRs
- new controller in the controller manager to handle the lifecycle of the ServiceCIDRs (do not leave orphan IPs)
- integration tests to cover the whole feature, since it is API based only integration will give us great coverage
- e2e tests
- `make update`


####

Demo https://gist.github.com/aojea/c20eb117bf1c1214f8bba26c495be9c7

#### TODO

Dual stackness and IP families 
```
kubectl apply -f https://gist.githubusercontent.com/aojea/90768935ab71cb31950b6a13078a7e92/raw/99ceac308f2b2658c7313198a39fbe24b155ae68/dual-stack.yaml
deployment.apps/httpd-deployment created
service/my-service-v4 created
service/my-service-prefer-dual created
Error from server (Invalid): error when creating "https://gist.githubusercontent.com/aojea/90768935ab71cb31950b6a13078a7e92/raw/99ceac308f2b2658c7313198a39fbe24b155ae68/dual-stack.yaml": Service "my-service-v6" is invalid: spec.ipFamilies[0]: Invalid value: "IPv6": not configured on this cluster
Error from server (Invalid): error when creating "https://gist.githubusercontent.com/aojea/90768935ab71cb31950b6a13078a7e92/raw/99ceac308f2b2658c7313198a39fbe24b155ae68/dual-stack.yaml": Service "my-service-require-dual" is invalid: spec.ipFamilyPolicy: Invalid value: "RequireDualStack": this cluster is not configured for dual-stack services
```

aojea committed:
kubectl describe

Change-Id: I0664e11a3a5549e1cc9602b22dcaf294200792a4

aojea committed:
Add ServiceCIDR to the REST Registry

Change-Id: If2ab86505b1ffd94100f575e682702679dd2da72

aojea committed:
API printers

Change-Id: I7a0777bbd0b10e1a849c7891439b00622ec6b09c

aojea committed:
e2e for ServiceCIDRs

Change-Id: Id7b711e30def4ac02f9e55425b1997d2cfe100f9

aojea committed:
service cidr controller manager

Controls the lifecycle of the ServiceCIDRs adding finalizers and
setting the Ready condition in status when they are created, and
removing the finalizers once it is safe to remove (no orphan IPAddresses)

An IPAddress is orphan if there are no ServiceCIDR containing it.

Change-Id: Icbe31e1ed8525fa04df3b741c8a817e5f2a49e80

aojea committed:
add ServiceCIDR APIs

Change-Id: Ia084c5505e43033ac34449031a1d32418ca326fd

Change-Id: Iafc236d456f7185a5c89a65d7b96245e04060013

aojea committed:
etcd data

Change-Id: I31c3a1c8fee7f3a681186b934bd7c6b36baee6df

aojea committed:
add apis to apiserver storage

Change-Id: I33dfbdad98695a6438c55d841139476cb1d740d7

aojea committed:
bootstrap controller for default ServiceCIDR

The bootstrap controller for ServiceCIDR ensures that the default
ServiceCIDR is created from the existing flags.

It follows the same behavior than the kubernetes.default Service,
it only creates the default ServiceCIDR if it doesn't exist, but
does not modify it despite the parameters doesn't match.

review: bootstrap controller for default ServiceCIDR

aojea committed:
add iptree radix tree util

A radix tree is required to optimize operations with subnets and
IP addresses.

Change-Id: I9fecc291efd39bdd7403c9675c047d6dee6018d2

Change-Id: I72c7bd5920a42bf35305443450c4ba97f857c492

aojea committed:
expand IPAllocator to work with multiple ServiceCIDRs

Create a new allocator that uses the ServiceCIDRs configured in the
system to create IPAllocators.

The CIDRAllocator will create IPAllocators per parent ServiceCIDRs,
since we allow overlapping, there is no need to have an allocator
per ServiceCIDR.
The benefit of the IPAllocator is that uses the informer cache as
storage, hence, it does not need to keep cache and as only as logical
abstraction. This allows to create and delete IPAllocators without
any penalty.

IPAllocators can allocate IP addresses only if they are ready (not
being deleted)

Change-Id: I3fdda69991907c39cca3120fe2d850f14dcccec2

aojea committed:
plumb the new multicidr clusterip allocator

aojea committed:
make repair loop multi service cidr aware

The repair loop controller watches the ServiceCIDRs configured
and use them to handle the repair of the IPAddresses assigned
by the kube-apiserver.

Change-Id: I8cfe8fd6285ea91192fc4ec72eaeea1eb004a235

Change-Id: If4be12e2c67b340d86c4efa2f9fb3672f0661636

aojea committed:
add integration test

Since ServiceCIDR and IPAddresses are mostly API driven integration
test will give us a good coverage, exercising real use cases like
the migration from one ServiceCIDR range to a new range.

aojea commented:
/assign @khenidak @thockin @smarterclayton 

This is the first phase of the KEP, it adds the IPAddress and ServiceCIDR object, it is https://github.com/kubernetes/kubernetes/pull/115075 +ServiceCIDR

Due to the complexity I prefer to add the controller logic to handle the ServiceCIDR lifecycle in a different PR, but this one allows people to start using the objects with its full functionality (see the integration test as example)

The 2 important changes are:
- The allocator uses IPAddresses instead of bitmaps, to remove the limitations imposed by the bitmap size.
- The allocator uses the ServiceCIDR object to configure multiple ranges.

I would love to have this PR in 1.27 so I can iterate, or it will be impossible to break this loop, my oldest branch is from 1.21 😅

aojea commented:
Sweeet

![image](https://user-images.githubusercontent.com/6450081/225010171-58f04712-e942-4468-8e7c-a5627dacb05d.png)

aojea commented:
this is postponed for 1.28

aojea committed:
fix IPAddress.Name validation message

Change-Id: Idc29166d69c1eadd6e4dac3bf16f4a6739f98be2

fedebongio commented:
/triage accepted

aojea commented:
/test

aojea commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

aojea commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

aojea commented:
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-e2e-gce-cos-alpha-features

aojea commented:
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-e2e-gce-cos-alpha-features

this should be the good one

aojea commented:
/assign @danwinship

danwinship commented on test/integration/servicecidr/migration_test.go:
```diff
@@ -0,0 +1,292 @@
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatalf("waiting for default service cidr ready condition set to true")
```
(so on all of these Poll error messages, you're discarding `err` even though the Poll returns an error that isn't "timed out" in some circumstances)

danwinship replied:
yeah, the Fatalf shoudl include the error

danwinship commented on test/integration/servicecidr/migration_test.go:
```diff
@@ -0,0 +1,292 @@
+		t.Fatalf("waiting for default service kubernetes.default to be migrated")
+	}
+
+	// The temporal ServiceCIDR can be deleted now since the Default ServiceCIDR will cover it
+	// Add a new service CIDR to be able to migrate the apiserver
+	if err := client2.NetworkingV1alpha1().ServiceCIDRs().Delete(context.Background(), "migration-cidr", metav1.DeleteOptions{}); err != nil {
```
second line of that comment doesn't seem like it belongs there?

danwinship commented on test/integration/servicecidr/servicecidr_test.go:
```diff
@@ -0,0 +1,345 @@
+		for _, s := range svcs.Items {
+			allIPs = append(allIPs, s.Spec.ClusterIP)
+		}
+		t.Fatalf("unexpected creation success. The following IPs exist: %#v. It should only be possible to allocate 2 IP addresses in this cluster.\n\n%#v", allIPs, svcs)
```
6, not 2

danwinship commented on test/integration/servicecidr/servicecidr_test.go:
```diff
@@ -0,0 +1,345 @@
+		t.Fatalf("got unexpected error: %v", err)
+	}
+	// wait the allocator process the new ServiceCIDR
+	time.Sleep(500 * time.Millisecond)
```
:raised_eyebrow:

aojea replied:
I need to remove this for the wait.Poll , thanks for cacthing it

danwinship commented on test/integration/servicecidr/servicecidr_test.go:
```diff
@@ -0,0 +1,345 @@
+// The test starts the apiserver with cidr1
+// Create Services to fill the range
+// Creates a new ServiceCIDR with the same range as cidr1
+// Deletes cidr1 since it is covered by the default ServiceCIDR
```
You are using "`cidr1`" here to mean both "the value `192.168.0.0/29`" and "the ServiceCIDR object named `cidr1`", which is confusing

danwinship commented on test/integration/servicecidr/servicecidr_test.go:
```diff
@@ -0,0 +1,345 @@
+// Creates a new ServiceCIDR with the same range as cidr1
+// Deletes cidr1 since it is covered by the default ServiceCIDR
+// Creates a new cidr2 with a different range than cidr1
+// Creates a new service so it picks an IPAddress on this range
```
"(because cidr1 is full)"

danwinship commented on test/integration/servicecidr/servicecidr_test.go:
```diff
@@ -0,0 +1,345 @@
+// Creates a new cidr3 that contains cidr2
+// Deletes cidr2 since it is covered by cidr3
+// Tries to delete cidr3 but is blocked since there is an IPAddress
+// Deletes the Service with the IPAddres blocking the deletion
```
IPAddres**s**

danwinship commented on test/integration/servicecidr/servicecidr_test.go:
```diff
@@ -0,0 +1,345 @@
+	ns := framework.CreateNamespaceOrDie(client, "test-service-cidr-deletion", t)
+	defer framework.DeleteNamespaceOrDie(client, ns, t)
+
+	// ServiceCIDR controller
```
maybe have a helper function for the shared setup code in these tests?

danwinship reviewed (commented):
(reviewing just the integration tests)

khenidak commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -523,6 +557,67 @@ func (r *RepairIPAddress) syncIPAddress(key string) error {
+		return err
+	}
+
+	treeV4 := iptree.New[string](false)
```
the question remains :-) do we really need to rebuild the entire tree everytime.

aojea replied:
the problem is that we allow overlapping and duplicate ranges and we can get in an inconsistent state if we process deltas , if I create ServiceCIDR A with range 192.168.0.0/24 and ServiceCIDR B with the same range, I will have to take into account that there is still another one if one of the subnets is deleted.
This is not different than what we are doing with the bitmap allocators, we always rebuild them from scratch, and. is much simpler than taking into account all the existing ranges, also, we don't expect ServiceCIDRs go beyond the 2 digits, right?

khenidak reviewed (commented):
Round 1: everything except allocator logic modification. Will look at them separately.

khenidak reviewed (commented):
Part 2: allocators.

khenidak commented:
I submitted a review, will do another read through next week. Thank you @aojea for this.

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -212,6 +242,9 @@ func (r *RepairIPAddress) RunUntil(onFirstSuccess func(), stopCh chan struct{})
 	onFirstSuccess()
 
+	// cidrWorker locks so we don't win anything parallelizing
+	go wait.Until(r.cidrWorker, r.workerLoopPeriod, stopCh)
```
"so we don't win anything by running multiple copies"?

You're still parallelizing it with ipWorker and svcWorker...

aojea replied:
yeah, but for handling the ServiceCIDR events we don't win anything running multiple workers, do we?

danwinship replied:
I was just saying that the comment is confusing. You say that it doesn't make sense to run it in parallel, but then you run it in parallel (with ipWorker and svcWorker)...

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -333,17 +366,18 @@ func (r *RepairIPAddress) syncService(key string) error {
+		r.muTree.Lock()
+		var ok bool
+		if family == netutils.IPv4 {
+			_, _, ok = r.treeV4.LongestPrefixMatch(clusterIP + "/32")
```
didn't you do this somewhere else too? `iptree.ContainsIP`?

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+	// are the ones with the longest prefix match.
+	var cidrs cidrSet
+	if ipFamily == string(v1.IPv4Protocol) {
+		_, cidrs, ok = c.treeV4.LongestPrefixMatch(ip.Name + "/32")
```
My earlier comment about .0 and broadcast addresses applies here and is more explicitly a bug; if you have allocators for `192.168.1.0/24` and `192.168.0.0/16`, then this will say that `192.168.1.0` belongs to `192.168.1.0/24`, but it actually belongs to `192.168.0.0/16`.

aojea replied:
I had a discussion with @thockin some time ago I can't find, but there was also an RFC explaining that  192.168.1.0 or 192.168.1.255 are a perfect valid address in 192.168.1.0/16 , meanwhile you don't use a broadcast medium as ethernet of course

danwinship replied:
Getting rid of the two unallocated IPs per CIDR seems like a great idea, if we think it's safe.

aojea replied:
nah, you are right , the IPAllocator does implement the same logic we always had, we don't allocate the .0 and the broadcast only in IPv6, is a very subtle bug, thanks for catching

aojea replied:
fixed

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+
+	// the only serviceCIDRs that can be affected by an IPAddress deletiong
```
"deletion"

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+			subnet, cidrs, ok := c.treeV4.ShortestPrefixMatch(cidr.Spec.IPv4)
+			// Check if there is other subnet that contains the existing one.
+			// The parent subnet can be greater or equal to the existing one.
+			if !ok || (subnet == cidr.Spec.IPv4 && cidrs.Len() < 2) {
```
Need to explain how this particular check implements the described logic, because it's not obvious

aojea replied:
fixed

aojea commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -2231,3 +2231,151 @@ func TestValidateIPAddressUpdate(t *testing.T) {
+					Name: "test-name",
+				},
+				Spec: networking.ServiceCIDRSpec{
+					IPv4: "192.168.0.1/24",
```
it has a .1 , it should be 192.168.0.0/24 ,

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+	defer c.muTree.Unlock()
+
+	// the only serviceCIDRs that can be affected by an IPAddress deletiong
+	// are the ones with the longest prefix match.
```
Hm... I'm having a hard time convincing myself that this is true for all combinations of overlapping ServiceCIDR Adds and Deletes, but I can't come up with an example where it fails...

aojea replied:
I put a lot of thought on all the edge cases and I still have that feeling that something is missing, what can I say 🤷

aojea replied:
ok, this has the issue you commented on the allocator, if CIDR1 has 192.168.0.0/16 and CIDR2 has 192.168.1.0/24, and there is only one IP 192.168.1.0, it should not block the deletion of CIDR2 but block the deletion of CIDR1

aojea replied:
fixed

aojea commented on test/integration/servicecidr/migration_test.go:
```diff
@@ -0,0 +1,292 @@
+	informers2.Start(ctx2.Done())
+
+	// delete the kubernetes.default service so the old DefaultServiceCIDR can be deleted
+	// and the new apiserver can take over
```
>  we could go an arbitrary amount of time without having a default ServiceCIDR.

There will be always a default ServiceCIDR, the thing is that if some admin wants to delete a ServiceCIDR, it has to create an overlapping ServiceCIDR so there are no orphan IPs. In this test, I choose a really complex scenario, sowe completely renumber the default ServiceCIDR and starts with a new range, that is why you need to delete it, but it can not be deleted because the kubernetes.default Service has an IP in that range.

> It seems like it would be better if the autocreated ServiceCIDR instead just had generateName: "kubernetes-" (or "default-" or whatever) rather than a specific hardcoded name, and there was a label or annotation to mark the ServiceCIDR that was currently the default...


That was the initial proposal from Tim, I went ahead and tried to implement it but I ended it discarding it and pushing back because it adds a lot of complexity in a critical path like the apiserver bootstrap and is also a problem that already exist,right now you can run apiserver with different flags and both will start but your Services will be complete broken https://github.com/kubernetes/kubernetes/issues/114743 . It is a kind of `sudo rm -rf /` operation too,

danwinship replied:
> > we could go an arbitrary amount of time without having a default ServiceCIDR.
> 
> There will be always a default ServiceCIDR

Well, there will always be a ServiceCIDR with the `DefaultServiceCIDRName`. But that ServiceCIDR doesn't always represent the `kube-apiserver --service-cluster-ip-range` value. And sometimes, _no_ ServiceCIDR represents the `kube-apiserver --service-cluster-ip-range` value.

aojea replied:
yes, that is on purpose, so we can make the behavior deterministic and match the existing for the "kubernetes.default" Service.
The unique name provides consistency and we define the object as the source of truth and the flags as the generator of the source of truth, we still need to flesh out more details before going to beta

aojea commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+		}
+	}
+
+	for _, cidr := range c.ipToCidrs(ip) {
```
1. golang already handles it https://go.dev/play/p/BNzyky43uOf
2. yes , overlapping ServiceCIDR is a requirement

aojea commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+	}
+	logger := klog.FromContext(ctx)
+	if c.queue.NumRequeues(key) < maxRetries {
+		logger.V(2).Info("Error syncing ServiceCIDR, retrying", "ServiceCIDR", key, "err", err)
```
the queue is ratelimited and the errors are tried with a backoff

aojea commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+	}()
+
+	// restore the radix tree from the current state
+	err := c.syncCIDRs()
```
we sync on ServiceCIDRs, IPs enqueue the ServiceCIDR they are directly depending on, the queue collapses the events related to the same object, I feel that your intuition is right and we can optimize, but I think this is easier to understand for the first iteration, I will add a TODO to verify if this presents some bottleneck

aojea commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+					return err
+				}
+			}
+			logger.V(4).Info("No IPv4 addresses referencing", "ServiceCIDR", key)
```
if there is any entry that belongs to the SErviceCIDR that loop exist, we only end here if the IPs reference the ServiceCIDR , that is what the message indicated

aojea commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	klog.FromContext(ctx).V(4).Info("Removed protection finalizer to ServiceCIDs", "ServiceCIDR", cidr.Name)
```
ServiceCIDRs are non/namespaced

aojea commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,219 @@
+
+	// default ServiceCIDR does not exist
+	klog.Infof("Creating default ServiceCIDR, ipv4: %s ipv6: %s", c.ipv4CIDR, c.ipv6CIDR)
+	serviceCIDR = &networkingapiv1alpha1.ServiceCIDR{
```
yes , is implicitly mentioned, the behavior of ServiceCIDR is the same as for the kubernetes.default Service, if it exists no apiserver  touches it, in this cse it will happen the same, the apiserver will not create the new one if there is still one created, the unique name provides us consistency

aojea commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,219 @@
+
+}
+
+func (c *Controller) syncStatus(serviceCIDR *networkingapiv1alpha1.ServiceCIDR) {
```
if is deleted, you recreate it first and this is called after

aojea commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,219 @@
+	}
+
+	// This controller will set the Ready condition to true if the Ready condition
+	// does not exist and the CIDR values match this controller CIDR values.
```
The controller manager may not be running to set the Status, and some projects start to use the apiserver standalone, I had the same thought as you, but I ended up with this becuase of the controller manager dependency

> one that as part of the Create instead

Status is a different subresource,  you can only set Spec or Status (AFAIK)

danwinship replied:
> Status is a different subresource, you can only set Spec or Status (AFAIK)

Weird, I think you're right but I had never noticed that that was true for Create rather than just Update...

aojea commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,219 @@
+			return
+		}
+	}
+	// set status to ready if the ServiceCIDR matches this configuration
```
it does not create the object

ibihim commented:
From sig-auth perspective the RBAC changes make sense.

If you need approval, ping a sig-auth member, once this PR is ready.

danwinship commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -808,3 +808,65 @@ func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList
+		}
+	}
+
+	return allErrs
```
Oh, I was thinking you could use a /32 but I guess that wouldn't work since it won't assign the first or last IP in a range... though we could special-case /32 and /128...

Anyway, the idea was that you could start the apiserver with a `--service-cluster-ip-range` that contained only a single IP, which would then get assigned to kubernetes.default, and then once it starts, you would create a ServiceCIDR which would define the CIDR to be used by every other Service.

This was an attempt to minimize the ugliness of the fact that Service CIDR configuration is now partially on the command-line and partially in API objects, by only using the command-line to configure the 1 Service IP that *can't* be configured by an API object. Although actually, that's not true anyway; you can connect to the apiserver via the node IP even before the service network is working. So I guess the better version of this proposal would be to allow starting kube-apiserver up without passing `--service-cluster-ip-range` at all, and it would just adopt the first ServiceCIDR it sees as the default. Or something?

aojea replied:
we have to keep the same validation we have already for the flag, that AFAIK only checks the max size https://github.com/kubernetes/kubernetes/blob/69c3b23abdbda53d14e14afc2af2bdfef23ac7b0/cmd/kube-apiserver/app/options/validation.go#L36, with this feature we remove that restriction, that is something that I realized that I missed

danwinship commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,219 @@
+func (c *Controller) syncStatus(serviceCIDR *networkingapiv1alpha1.ServiceCIDR) {
+	// syncStatus is only called if the serviceCIDR exists or was just created
+	// so we can close the readyCh to indicate the apiserver is healthy
+	select {
+	case <-c.readyCh:
+	default:
+		close(c.readyCh)
+	}
```
I mean, I know it _works_, but is that the sort of thing we do elsewhere? It just seems a little confusing.

aojea replied:
it is used in the golang library, that is from where I borrowed

aojea commented on pkg/registry/networking/servicecidr/storage/storage.go:
```diff
@@ -0,0 +1,101 @@
+/*
```
boilerplate, I don't know either

aojea commented on pkg/util/iptree/iptree.go:
@thockin WDYT?

aojea commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -157,3 +157,77 @@ type IPAddressList struct {
+	ServiceCIDRConditionReady = "Ready"
+	// NetworkConflict represents a reason where the ServiceCIDR can not be used
+	// because there is overlapping with other networks in the cluster.
+	ServiceCIDRReasonNetworkConflict = "NetworkConflict"
```
yes, this is a leftover of the previous implementation when we wanted to model the whole network and part of the internet XD

aojea commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -157,3 +157,77 @@ type IPAddressList struct {
+	ServiceCIDRReasonNetworkConflict = "NetworkConflict"
+	// ServiceCIDRReasonOrphanIPAddress represents a reason where a ServiceCIDR can not be deleted because
+	// it will leave orphan IP addresses.
+	ServiceCIDRReasonOrphanIPAddress = "OrphanIPAddress"
```
yes, I want to give clear indications to users on why things are happening, so I think Reason should be `Terminating` and then the message explain that is "blocked becuse there are IPs depending on this serviceCIDRs" , WDYT?

danwinship replied:
sure

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+
+// Insert the specified subnet and the value associated.
+// Retyrns true if updated.
+func (t *Tree[T]) Insert(subnet string, v T) (T, bool) {
```
it borrows the semantics from https://github.com/armon/go-radix, the tree implementation is based on that one

danwinship replied:
That implementation always returns either `old, true` or `nil, false`, which makes more sense. It doesn't have the hard-to-explain `v, false` case.

aojea replied:
I can make the return value a pointer *T

danwinship replied:
No, it's not about that. I mean that line 301 should be `return zeroT, false` rather than `return v, false`. Or if not, then you _really_ need to update the documentation to explain why it makes sense to return `v` rather than `zeroT` only in that one case.

aojea replied:
I think that returning the value does not make much sense and is hard to think about in this case, will change the API to just return true if update

aojea commented on pkg/util/iptree/iptree_test.go:
```diff
@@ -0,0 +1,540 @@
+	}
+}
+
+func generateRandomCIDR(is6 bool, number int) sets.Set[string] {
```
It was a long time ago I did this, I think it was because of performance issues, but I can't say now

aojea commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -163,6 +166,32 @@ func NewRepairIPAddress(interval time.Duration,
+	// ServiceCIDR has to rebuild the state from the informer cache and not process events
+	// directly because ServiceCIDR can contain the same CIDRs, that means that is impossible
+	// to handle with the radix tree, since the key is unique.
+	_, _ = serviceCIDRInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```
yes, I had that in the handlers before XD and then the order of events caused problems

aojea replied:
the comment is not clarifying nothing, will remove

aojea commented on pkg/apis/networking/types.go:
```diff
@@ -695,9 +694,6 @@ type ParentReference struct {
 	Name string
-	// UID is the uid of the object being referenced.
-	// +optional
-	UID types.UID
```
ok

aojea replied:
done https://github.com/kubernetes/kubernetes/pull/121106

aojea commented on pkg/apis/networking/types.go:
```diff
@@ -711,3 +707,57 @@ type IPAddressList struct {
+// ServiceCIDR defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::/64).
+// This range is used by the cluster to allocate the ClusterIPs associated to the Services object.
+type ServiceCIDR struct {
+	metav1.TypeMeta
```
See line 712, are nonNamespaces

> // +genclient:nonNamespaced

aojea commented on pkg/apis/networking/types.go:
```diff
@@ -711,3 +707,57 @@ type IPAddressList struct {
+	Status ServiceCIDRStatus
+}
+
+// ServiceCIDRSpec describe how the ServiceCIDR's specification looks like.
```
this is only for Services, ServiceCIDRs and IPAddresses, after all the discussions where we tried to get a more comprehensible definition of the cluster network we settled on , kubernetes only owns the Service CIDRs and the rest is left to implementations, actually we dropped ClusterCIDR because of this

aojea commented on pkg/apis/networking/fuzzer/fuzzer.go:
```diff
@@ -108,3 +115,22 @@ func generateRandomIP(is6 bool, c fuzz.Continue) string {
+		if is6 {
+			return "2001:db8::/64"
+		}
+		return "192.168.1.0/24"
```
most of these things are to avoid linter complains, will add the panics

aojea commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -808,3 +808,65 @@ func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv6"), cidrConfig.Spec.IPv6, "wrong CIDR format, IP doesn't match network IP address"))
+			}
+			if prefix.Masked().String() != cidrConfig.Spec.IPv6 {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv6"), cidrConfig.Spec.IPv6, "CIDR not in canonical format"))
```
I didn't see  places that validation add variable content to the message

danwinship replied:
Hm... that's true. It just seems a little rude; the user might not understand what "canonical format" means, and googling "cidr canonical format" doesn't suggest anything helpful.

aojea replied:
there is an rfc or something I remember we added in other validation let me check

aojea replied:
for IPv6 there is an rfc, for IPv4 I found also libraries implementing the canonicalization https://docs.servicenow.com/en-US/bundle/vancouver-api-reference/page/app-store/dev_portal/API_reference/IPAddress/concept/IPAddressAPIBoth.html

aojea commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -2231,3 +2231,151 @@ func TestValidateIPAddressUpdate(t *testing.T) {
+				},
+			},
+		},
+	}
```
added

aojea commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -2231,3 +2231,151 @@ func TestValidateIPAddressUpdate(t *testing.T) {
+				return out
+			}, expectErr: true,
+		},
+	}
```
this tests the validation logic, we don't have validation logic for the status

aojea commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,219 @@
+	// This controller will set the Ready condition to true if the Ready condition
+	// does not exist and the CIDR values match this controller CIDR values.
+	for _, condition := range serviceCIDR.Status.Conditions {
+		if condition.Type == networkingapiv1alpha1.ServiceCIDRConditionReady {
```
yeah, but I want to get the condtion Reason and Message if it does not exist and is only used in one place, so I don't think adds much value

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+}
+
+func (n *node[T]) mergeChild() {
+	// can not merge if there are two or no childs
```
is a leaf if there are no children, right?

danwinship replied:
But it can also be a leaf even if there are children, can't it? (Which is a weird definition of "leaf", but whatevs.)

e.g., the check at line 461 assumes that `parent` can both have a child and be a leaf (and that it shouldn't be merged if it's a leaf).

aojea replied:
leaf is confusing indeed, it really means "node that can be a leaf because it contains a value" or "node that is not added by the user"

aojea replied:
I will switch it to public , as it is a node that is set by the user

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+}
+
+// New creates a new Radix Tree for IP addresses
+func New[T any](is6 bool) *Tree[T] {
```
I don't want to get the IPFamily type here because it maybe reused and this will bring the kubernetes type

aojea replied:
I'm going to make the Tree to be able to handle both IPv4 and IPv6 it simplifies its usage a lot

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+	n := t.root
+	// bit position is given by the mask bits
+	bitPosition := 0
+	// mask the address
+	address := prefix.Masked().Addr()
+	mask := prefix.Bits()
+	for bitPosition <= mask {
```
I prefer to sanitize all the variables anyway , I can't remember now and most probably you are right, but I had some subtle bugs related to sanitizing so most probably I'm being extra paranoid

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+	return p.String(), v, ok
+}
+
+func (t *Tree[T]) shortestPrefix(prefix netip.Prefix) (netip.Prefix, T, bool) {
```
These is the WET and DRY thing, let's keep it separate by now

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,413 @@
+func (c *CIDRAllocator) EnableMetrics() {}
+
+// DryRun returns a random allocator
+func (c *CIDRAllocator) DryRun() Interface {
```
this is solved just by returning the DryRun() as you comment in https://github.com/kubernetes/kubernetes/pull/116516/commits/389fe36dd6094a1ab62b2f67b5d3973a2d7fd810#r1349337220, good catch

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,413 @@
+
+// Allocator implements current ipallocator interface using
+// ServiceCIDR and IPAddress API objects
+type CIDRAllocator struct {
```
Allocator and MultiAllocator sounds coherent,

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,413 @@
+
+// Allocator implements current ipallocator interface using
+// ServiceCIDR and IPAddress API objects
+type CIDRAllocator struct {
```
Allocator and MultiAllocator sounds coherent,

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,413 @@
+	family := api.IPv4Protocol
+	if isIPv6 {
+		family = api.IPv6Protocol
+	}
```
this has to be revisited later as it depends on the allocator inicializaation , right now the allocators set the default family based on the service-cidr flag, but we are going to make the allocators configurable, so it may happen an apiserver can switch the default ip family ... I´ll add a TODO to not forget

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,413 @@
+	}
+	key := fmt.Sprintf("%s/%d", ip.String(), mask)
+	// Use the largest subnet to allocate addresses
+	_, allocator, ok := c.tree.ShortestPrefixMatch(key)
```
added an unit test just for this specific case

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,413 @@
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+
+	// TODO add strategy to return the allocator with more free addresses
+	for _, allocator := range c.tree.Parents() {
```
the allocators randomize the IPs assigned, I don't think that deterministic is a good strategy, what I'm trying to mean is that it should be weighted, i.e. if I have 3 allocators with 256, 1024 and 4096 addresses each, we should make first allocator less likely to hit

aojea replied:
however, I can see how this can change in the future, so we may add an Strategy like we added in the bitmap allocators, I will add a TODO to revisit on the next iteration

aojea replied:
This will need to be revisited before beta

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,413 @@
+	return nil, ErrFull
+}
+
+func (c *CIDRAllocator) AllocateNext() (net.IP, error) {
```
baby steps, we´ll get there but this is the safest way

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator_test.go:
```diff
@@ -0,0 +1,416 @@
+	ipInformer := informerFactory.Networking().V1alpha1().IPAddresses()
+	ipStore := ipInformer.Informer().GetIndexer()
+
+	client.PrependReactor("create", "servicecidrs", k8stesting.ReactionFunc(func(action k8stesting.Action) (bool, runtime.Object, error) {
```
if I delete the reactors the test fail, AFAIK the fakeClient does not modify the informer storage

aojea commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -333,17 +366,18 @@ func (r *RepairIPAddress) syncService(key string) error {
+		r.muTree.Lock()
+		var ok bool
+		if family == netutils.IPv4 {
+			_, _, ok = r.treeV4.LongestPrefixMatch(clusterIP + "/32")
```
I still need to build the prefix for the IP/32 or IP/128, however, this code can be refactored to make better as you say, I will add a TODO

aojea commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+// Controller manages selector-based service ipAddress.
+type Controller struct {
+	client           clientset.Interface
+	eventBroadcaster record.EventBroadcaster
```
it is used, `		c.eventRecorder.Eventf(cidr, v1.EventTypeWarning, "KubernetesServiceCIDRError", "The ServiceCIDR Status can not be set to Ready=True")
`

aojea commented on test/integration/servicecidr/migration_test.go:
```diff
@@ -0,0 +1,292 @@
+	}
+
+	// start second apiserver with the new range and new service cidr controller
+	s2 := kubeapiservertesting.StartTestServerOrDie(t,
```
the DefaultServiceCIDR only is recreated if is deleted

aojea commented on test/integration/servicecidr/servicecidr_test.go:
```diff
@@ -0,0 +1,345 @@
+limitations under the License.
+*/
+
+package servicecidr
```
that is covered on the unit tests, integration just assert the most complex behaviors, I don't think we need to keep revalidating all those details, since we are properly implementing a pyramid of testing

aojea commented on test/integration/servicecidr/migration_test.go:
```diff
@@ -0,0 +1,292 @@
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
```
https://github.com/kubernetes/kubernetes/pull/116516/commits/7ea187a5ca4e6d6587c619f99cc4503446cbe875#r1360472107

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+limitations under the License.
+*/
+
+package iptree
```
if you search around most of the trees are named this way

https://github.com/zmap/go-iptree
https://pkg.go.dev/github.com/infobloxopen/go-trees/iptree

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+}
+
+// Get returns true if the subnet exists in the tree
+func (t *Tree[T]) Get(subnet string) (T, bool) {
```
do we need that?  I prefer to not do validation errors here, if is invalid is just that does not exist

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+	} else {
+		return 0
+	}
+
+}
```
fixed

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+
+	m1 := t.ToMap()
+	m2 := tree.ToMap()
+	return reflect.DeepEqual(m1, m2)
```
yeah, removed this method as it is not used anywhere

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+}
+
+// Parents return the Parents nodes in a tree
+func (t *Tree[T]) Parents() map[string]T {
```
done

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+type WalkFn[T any] func(s netip.Prefix, v T) bool
+
+// Walk is used to walk the tree
+func (t *Tree[T]) Walk(fn WalkFn[T]) {
```
done

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+	// Delete the value
+	n.leaf = false
+	val := n.val
+	t.size--
```
the t.root is special, we can not merge it, I'm adding more comments to provide more details

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+		parent = n
+		n = n.child[childIndex]
+
+		// No child, create one
+		if n == nil {
```
let me change it, you are right

aojea committed:
make update

Change-Id: I4007d4f6dda24b5c74f602d35191556f315de222

aojea commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -808,3 +808,65 @@ func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList
+			if prefix.Addr() != prefix.Masked().Addr() {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv6"), cidrConfig.Spec.IPv6, "wrong CIDR format, IP doesn't match network IP address"))
+			}
+			if prefix.Masked().String() != cidrConfig.Spec.IPv6 {
```
fixed, this was hard to understand 😄 , thanks

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,602 @@
+			break
+		}
+
+	}
```
I refactored all the loops with your suggestion, this reads much better now

aojea commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,500 @@
+	}
+
+	// Deleting ....
+	if !cidr.GetDeletionTimestamp().IsZero() {
```
done

aojea commented:
@danwinship @khenidak I think I addressed all the comments, the most brittle parts were the iptree and the servicecidr controller in the controller manager.

I refactored the iptree with the suggestions, changed the API to make it simpler to use on Kubernetes, and made it IP family agnostic so we don't have to be using if IPv6 do treeV6 or if IPv4 do tree IPv4.

The serviceCidr controller in the controller manager is super tricky, I split the logic in smaller functions and added a bunch of unit tests

PTAL, since the PR is super huge and we are getting closer to Kubecon and code freeze I appreciate if you comment on the reviews things that need to be addressed in the PR and things I should add a TODO to follow up, since this is still in alpha.

aojea commented:
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-e2e-gce-cos-alpha-features

aojea commented:
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-e2e-gce-cos-alpha-features

aojea commented:
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-e2e-gce-cos-alpha-features

/retest
Fakes are  https://github.com/kubernetes/kubernetes/issues/121313 and https://github.com/kubernetes/kubernetes/issues/121209

aojea commented:
Ok, the kind job for alpha features is broken also in others PR https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121105/pull-kubernetes-e2e-kind-alpha-features/1714601745411215360 , I have to investigate that

The other failures are not important, one is because the cluster does not have the apis enabled and tries.to run the e2e tests

This is ready for review

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,641 @@
+
+type cidrSet struct {
+	sets.Set[string]
+}
```
why do you need this? can't you just say that tree is an `*iptree.Tree[sets.Set[string]]`? Or at least just do `type cidrSet sets.Set[string]` rather than wrapping it in a struct?

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,641 @@
+// cidrToCidrs, given a ServiceCIDR return the ServiceCIDRs that contain or are contained
+// this is required because adding or removing a CIDR will require to recompute the
+// state of each ServiceCIDR to check if can be unblocked on deletion.
+func (c *Controller) cidrToCidrs(cidr *networkingapiv1alpha1.ServiceCIDR) []string {
```
Again, confusing name: I would have assumed this was about CIDR strings, not ServiceCIDR object names. `overlappingServiceCIDRs`?

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,641 @@
+
+// ipToCidrs, given an IPAddress return the ServiceCIDRs that contains the IP,
+// as it may block or be blocking the deletion of the ServiceCIDRs that contain it.
+func (c *Controller) ipToCidrs(ip *networkingapiv1alpha1.IPAddress) []string {
```
`containingServiceCIDRs`?

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,641 @@
+
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+	cidrs := []string{}
```
Naming comment applies to variables too. I have a strong expectation that "cidr" refers to CIDR strings (or `*net.IPNet`s or `netip.Prefix`es), not ServiceCIDR object names.

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,641 @@
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	klog.FromContext(ctx).V(4).Info("Removed protection finalizer to ServiceCIDs", "ServiceCIDR", cidr.Name)
```
typo "ServiceCIDs"
also, "from", not "to"

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,641 @@
+func ipAddressToPrefix(ip *networkingapiv1alpha1.IPAddress) (netip.Prefix, error) {
+	labelFamily, ok := ip.Labels[networkingapiv1alpha1.LabelIPAddressFamily]
+	if !ok {
+		return netip.Prefix{}, fmt.Errorf("IPAddress without required label: %s", networkingapiv1alpha1.LabelIPAddressFamily)
```
Given that checking the label introduces another "can't happen, but..." situation, why not just check `netutils.IsIPv6String()`?

aojea replied:
because sometimes things that can not happen , just happen and this time we know it 😄

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,641 @@
+// but is not the network address and also, if IPv4, is not the broadcast address.
+// This is required because the Kubernetes allocators reserve these addresses
+// so IPAddresses can not block deletion of this ranges.
+func prefixContainIP(prefix netip.Prefix, ip netip.Addr) bool {
```
As noted above the problem still exists in other places. Basically _everywhere_ you look up an IP in an iptree.

aojea replied:
I moved it as a method of the tree, does not make sense to replicate everywhere are you pointed out

khenidak commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,592 @@
+
+// assume starts at 0 from the MSB: 0.1.2......31
+// return 0 or 1
+func getBitFromAddr(ip netip.Addr, pos int) int {
```
I think we can simplify this.
1. Normalize `pos` to zero based pos (no need to add one by caller and deducted here).
2. Preflight boundaries: `if(pos < 0 || !(pos < len(slice)))`.
3. Create a mask separately `mask := 1 << (pos % 32)`
4. Get an int by its position and mask it `return Binary.LittleEndian(slice[pos/32:8]) & mask`

aojea replied:
ipv6 addresses is the problem, so we can get a 4 or 16 bytes array.
To share the same code we calculate first which byte in the array contains the pos we want, and then we mask that byte.

khenidak reviewed (commented):
Still looking at tree implementation. But here is what i have so far.

aojea commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -721,3 +721,66 @@ func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList
+	allErrs := apivalidation.ValidateObjectMeta(&cidrConfig.ObjectMeta, false, ValidateServiceCIDRName, field.NewPath("metadata"))
+
+	if cidrConfig.Spec.IPv4 == "" && cidrConfig.Spec.IPv6 == "" {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("Spec"), cidrConfig.Spec, "at least one CIDR required"))
```
During the next phase we'll have to deal with the problem of the primary IPFamily, we have a notion of primary IP family and the apiservers has to be consistent, since when you create a Service and you don't specify a ClusterIP, it will assign an IP from the primary ... if we do the labels there can be races, the single Name uniqueness iprovide us consistency across all apiservers (that is the trick we are doing with the IPAddresses names) ... we'll have to figure out how to indicate IPv4,IPv6 or IPv6,IPv4 but we can use an annotation for that.

The slice of strings has two drawbacks in my opinion, if we do that we better move to a Singleton object and we allow Updates, but is easier for the user to make mistakes and is harder to implement the orphaning detection logic

aojea replied:
There can be only one primary IP family and it has to be the same across the cluster and you can not cross validate, so having different objects with different primary IP families is not valid, the same applies to arrays, you need to apply a convention  that is not immediate to know from the API object

From the user POV I think that is easier to understand having fields IPv4 and IPv6 , also if you think in the userspace tools, you already have this distinction ,  `ping` and `ping6` , `iptables` and `ip6tables` ...

danwinship replied:
> you need to apply a convention that is not immediate to know from the API object

Ah, true. But that just means that `primaryCIDR`/`secondaryCIDR` isn't an option here. `[]CIDRs` still is.

> From the user POV I think that is easier to understand having fields IPv4 and IPv6

If you have a dual-stack cluster, then I'd say it's equally easy to understand either way. We already have lots of fields that mix IPv4 and IPv6 IPs together (`podIPs`, `clusterIPs`, `node.status.addresses`) and nobody gets confused by them.

If you _don't_ have a dual-stack cluster, then I think explicit v4/v6 fields are actually harder to understand. Especially, it is harder to document; rather than saying "set `CIDR` to what you want the new CIDR to be", you have to say "if you're in a single-stack IPv4 cluster, then set `IPv4` to what you want the new CIDR to be, and if you're in a single-stack IPv6 cluster, then set `IPv6` to what you want the new CIDR to be."

aojea replied:
Discussed with Dan offline, for reference, after this PR we still have to solve the problem of defaulting the Primary IPFamily across the cluster, using an Array solves this problem because we can follow the same convention we are using now with the flags.

Since this is alpha, all code is gated and adding the logic to solve the primary ip family in this PR will increase a lot the complexity of the review, we defer this discussion for the next iteration, including the change of the fields by an array

aojea replied:
@danwinship @khenidak @thockin please comment on the doc https://docs.google.com/document/d/1-CKDPRHsG6XfnGs-Bb7ptWgDtTq9-hj8aJQpGz1CSh8/edit , I think Dan is right and an array limited to only two cidrs max (one IPv4 and other IPv6) is the best alternative for the API. If you can agree today I will update the PR and the KEP this weekend (code freeze is next wednesday)

aojea commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,223 @@
+
+func (c *Controller) syncStatus(serviceCIDR *networkingapiv1alpha1.ServiceCIDR) {
+
+	// don's sync the status of the ServiceCIDR if is being deleted,
```
my bad, it was late so I may did that inadvertently, sorry

aojea commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller_test.go:
```diff
@@ -0,0 +1,186 @@
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:              DefaultServiceCIDRName,
+						DeletionTimestamp: &now,
```
nice tip, that is much better than declaring it outside

aojea commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller_test.go:
```diff
@@ -0,0 +1,186 @@
+		t.Run(tc.name, func(t *testing.T) {
+			client, controller := newController(t, tc.cidrs)
+			controller.sync()
+			expectAction(t, client.Actions(), tc.actions)
+
+		})
```
:/ I have to check my linter config, I did some changes recently

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,592 @@
+type WalkFn[T any] func(s netip.Prefix, v T) bool
+
+// DepthFirstWalk is used to walk the tree of the corresponding IP family
+func (t *Tree[T]) DepthFirstWalk(isIPv6 bool, fn WalkFn[T]) {
```
I was going to do that but the v1.IPFamily is an enum with 3 values, this tree is IPv4 or IPv6, it can not be other thing, and create a yet another new enum just for this :(

thockin replied:
The third option being "" ?  Just return immediately - you found no IPs of family ""

aojea commented on pkg/util/iptree/iptree_test.go:
```diff
@@ -0,0 +1,585 @@
+	"github.com/google/go-cmp/cmp"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
```
I though tthe same , it is in 

```
coverage: 88.5% of statements
```

and you can see that there is also fuzzy tests and benchmarks

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,592 @@
+	// walk the network bits of the prefix
+	for bitPosition < mask {
+		// Look for a child checking the bit position after the mask
+		n = n.child[getBitFromAddr(address, bitPosition+1)]
```
so we are in pos X in the tree, and this is binary so we either go right or left, and the X+1 pos will tell us which one

aojea commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,223 @@
+	}
+	// we construct our own informer because we need such a small subset of the information available.
+	// The kubernetes.default service
+	c.serviceCIDRInformer = networkingv1alpha1informers.NewFilteredServiceCIDRInformer(client, 12*time.Hour,
```
this is coming from the controlplane instantiation that has access to all shared informers, but yeah, you have to have that previous knowledge otherwise the comment does not have any sense

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,592 @@
+	return false
+}
+
+// AllPrefixes is used to walk the corresponding IP family tree and convert it into a map
```
I'll remove it

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,592 @@
+}
+
+// WalkPrefix is used to walk the tree under a prefix
+func (t *Tree[T]) WalkPrefix(prefix netip.Prefix, fn WalkFn[T]) {
```
all of them "Walk" the tree, once from root to the prefix, other from the prefix to the lead and other the entire tree using the DepthFirst algorithm, I think  that Walk is overloaded but from the description sounds it is understandable what each of them use, I will update the docs to make this more clear

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,592 @@
+}
+
+// WalkPath is used to walk the tree, but only visiting nodes
+// from the root down to a given public. Where WalkPrefix walks
```
the thing is that the nodes with value are the ones that are really public, are those the users can modify, the others type of nodes are internal created to build the tree, but users can not access them. I think that public as in the context of variables public or private fits better. Also, users don't have to set a value, so not having value does not really represents it

aojea replied:
it is not really to a value, is to a give IP prefix

aojea commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,592 @@
+	child [2]*node[T] // binary tree
+}
+
+func (n *node[T]) isPublic() bool {
```
removed

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,441 @@
+
+// NewMetaAllocator returns an IP allocator associated to a network range
+// that use the IPAddress objects to track the assigned IP addresses,
+// using an informer cache as storage.
```
yes

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,441 @@
+
+		if c.ipFamily == api.IPv4Protocol &&
+			cidr.Spec.IPv4 != "" {
+			cidrsSet.Insert(cidr.Spec.IPv4)
```
I think that before I had something else or I was too asleep, multiline here does not make sense, as you say. fixed

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,441 @@
+	prefix, err := netip.ParsePrefix(ip.String() + mask)
+	if err != nil {
+		return nil, err
+	}
```
I just found this is already documented https://pkg.go.dev/net/netip#AddrFromSlice

aojea replied:
I want to assert on the allocator family not in the passed IP, I have to revisit this to see if we can make it ip family agnostic

aojea replied:
but I rather postpone it to a next phase, when we deal with the primary ip family thing, otherwise this will be make this PR more complicated

aojea replied:
I love your suggestion though, changed to 

```
	address, ok := netip.AddrFromSlice(ip)
	if !ok {
		return nil, fmt.Errorf("can not parse IP %s", ip.String())
	}
	var prefix netip.Prefix
	if c.ipFamily == api.IPv4Protocol {
		prefix = netip.PrefixFrom(address, 32)
	} else {
		prefix = netip.PrefixFrom(address, 128)
	}

```

aojea replied:
great :/

```
DEBUG  ::ffff:192.168.0.1/32 192.168.0.1
```

aojea replied:
ok, this makes it
```
// ipToAddr converts a net.IP to a netip.Addr
// if the net.IP is not valid it returns an empty netip.Addr{}
func ipToAddr(ip net.IP) netip.Addr {
	// https://pkg.go.dev/net/netip#AddrFromSlice can return an IPv4 in IPv6 format
	// so we have to check the IP family to return exactly the format that we want
	// address, _ := netip.AddrFromSlice(net.ParseIP(192.168.0.1)) returns
	// an address like ::ffff:192.168.0.1/32
	bytes := ip.To4()
	if bytes == nil {
		bytes = ip.To16()
	}
	address, _ := netip.AddrFromSlice(bytes)
	return address
}

```

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator_test.go:
```diff
@@ -0,0 +1,487 @@
+	ipInformer := informerFactory.Networking().V1alpha1().IPAddresses()
+	ipStore := ipInformer.Informer().GetIndexer()
+
+	client.PrependReactor("create", "servicecidrs", k8stesting.ReactionFunc(func(action k8stesting.Action) (bool, runtime.Object, error) {
```
it does not work for me with `TestCIDRAllocateMultiple`

aojea replied:
```
rep -r PrependReact pkg/controller/
pkg/controller/endpointslice/endpointslice_controller_test.go:  client.PrependReactor("create", "endpointslices", k8stesting.ReactionFunc(func(action k8stesting.Action) (bool, runtime.Object, error) {
pkg/controller/endpointslice/endpointslice_controller_test.go:  client.PrependReactor("update", "endpointslices", k8stesting.ReactionFunc(func(action k8stesting.Action) (bool, runtime.Object, error) {
pkg/controller/history/controller_history_test.go:                      client.PrependReactor("*", "*", test.reactor)
pkg/controller/podautoscaler/horizontal_test.go:        testScaleClient.PrependReactor("get", "replicationcontrollers", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:        testScaleClient.PrependReactor("get", "replicationcontrollers", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:        testClient.PrependReactor("list", "pods", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:                testMetricsClient.PrependReactor("list", "pods", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:                testCMClient.PrependReactor("get", "*", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:                testEMClient.PrependReactor("list", "*", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:        testScaleClient.PrependReactor("get", "replicationcontrollers", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:        testScaleClient.PrependReactor("update", "replicationcontrollers", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:        testClient.PrependReactor("list", "horizontalpodautoscalers", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:        testClient.PrependReactor("update", "horizontalpodautoscalers", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:        testEMClient.PrependReactor("list", "*", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/podautoscaler/horizontal_test.go:        testEMClient.PrependReactor("list", "*", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/endpointslicemirroring/utils_test.go:    client.PrependReactor("create", "endpointslices", k8stesting.ReactionFunc(func(action k8stesting.Action) (bool, runtime.Object, error) {
pkg/controller/endpointslicemirroring/utils_test.go:    client.PrependReactor("update", "endpointslices", k8stesting.ReactionFunc(func(action k8stesting.Action) (bool, runtime.Object, error) {
pkg/controller/replicaset/replica_set_test.go:  client.PrependReactor("update", "replicasets",
pkg/controller/serviceaccount/tokens_controller_test.go:                                client.Fake.PrependReactor(reactor.verb, reactor.resource, reactor.reactor(t))
pkg/controller/resourceclaim/controller_test.go:                                fakeKubeClient.PrependReactor("create", "resourceclaims", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/resourceclaim/controller_test.go:        fakeClient.PrependReactor("create", "resourceclaims", createResourceClaimReactor())
pkg/controller/resourceclaim/controller_test.go:// the fake client. Add it with client.PrependReactor to your fake client.
pkg/controller/namespace/deletion/namespaced_resources_deleter_test.go: mockMetadataClient.PrependReactor("delete-collection", "flakes", ns1FlakesNotFound)
pkg/controller/namespace/deletion/namespaced_resources_deleter_test.go: mockMetadataClient.PrependReactor("list", "flakes", ns1FlakesNotFound)
pkg/controller/volume/attachdetach/statusupdater/node_status_updater_test.go:   fakeKubeClient.PrependReactor("patch", "nodes", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/volume/pvcprotection/pvc_protection_controller_test.go:                  client.Fake.PrependReactor(reactor.verb, reactor.resource, reactor.reactorfn)
pkg/controller/volume/pvprotection/pv_protection_controller_test.go:                    client.Fake.PrependReactor(reactor.verb, reactor.resource, reactor.reactorfn)
pkg/controller/volume/ephemeral/controller_test.go:                             fakeKubeClient.PrependReactor("create", "persistentvolumeclaims", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/daemon/daemon_controller_test.go:                clientset.PrependReactor("update", "daemonsets", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/daemon/daemon_controller_test.go:                clientset.PrependReactor("update", "daemonsets", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/daemon/daemon_controller_test.go:                                clientset.PrependReactor("update", "daemonsets", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/daemon/daemon_controller_test.go:                clientset.PrependReactor("update", "daemonsets", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/daemon/daemon_controller_test.go:                clientset.PrependReactor("update", "daemonsets", func(action core.Action) (handled bool, ret runtime.Object, err error) {
pkg/controller/disruption/disruption_test.go:   dc.coreClient.Fake.PrependReactor("update", "poddisruptionbudgets", func(a core.Action) (handled bool, obj runtime.Object, err error) {
pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:                         item.fakeNodeHandler.Clientset.PrependReactor("update", "pods", item.updateReactor)
pkg/controller/statefulset/stateful_set_test.go:                ssc.kubeClient.(*fake.Clientset).PrependReactor("patch", "pods", func(action core.Action) (bool, runtime.Object, error) {
pkg/controller/statefulset/stateful_pod_control_test.go:        fakeClient.PrependReactor("update", "pods", func(action core.Action) (bool, runtime.Object, error) {
pkg/controller/statefulset/stateful_pod_control_test.go:        fakeClient.PrependReactor("update", "pods", func(action core.Action) (bool, runtime.Object, error) {
pkg/controller/clusterroleaggregation/clusterroleaggregation_controller_test.go:             
```

aojea replied:
I hate the fake client though, I don't trust it to do the right thing, I have been hit a few times

aojea commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -330,20 +359,28 @@ func (r *RepairIPAddress) syncService(key string) error {
+		prefix, err := netip.ParsePrefix(clusterIP + mask)
+		if err != nil {
+			// ClusterIP is corrupt, ClusterIPs are already validated, but double checking here
+			// in case there are some inconsistencies with the parsers
```
yes, using the helper and your previous suggestion looks much better

aojea commented on pkg/registry/core/rest/storage_core.go:
```diff
@@ -345,7 +346,16 @@ func (c *Config) newServiceIPAllocators() (registries rangeRegistries, primaryCl
+		// TODO(aojea) Revisit the initialization of the allocators
+		// since right now it depends on the service-cidr flags and
+		// sets the default IPFamily that may not be coherent with the
+		// existing default ServiceCIDR
```
that's the goal yes, commented before, need to deal with the defaulting,  next phase will make it

aojea commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -330,20 +359,28 @@ func (r *RepairIPAddress) syncService(key string) error {
-			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]: %s is not within the configured Service CIDR %s; please recreate service", v1Family, ip, network.String())
-			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s is not within the service CIDR %s; please recreate", v1Family, ip, svc.Namespace, svc.Name, network.String()))
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]: %s is not within the configured Service CIDR; please recreate service", v1Family, ip)
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s is not within any service CIDR; please recreate", v1Family, ip, svc.Namespace, svc.Name))
```
fixed

aojea commented:
Changelog:
- addressed comments
- iptree now has a new method that give an IP address it returns the list of prefixes that contains the IP as host, iu.e. no network address or broadcast for ipv4
- modified all the controllers to consume the new iptree method, it simplifies the code

aojea commented:
> Job execution failed: Pod got deleted unexpectedly

/test pull-kubernetes-node-e2e-containerd

aojea commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

danwinship commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,684 @@
+// or, if IPv4, the broadcast address (RFC 1878).
+func (t *Tree[T]) GetHostIPPrefixMatches(ip netip.Addr) map[netip.Prefix]T {
+	// walk the tree to find all the prefixes containing this IP
+	var ipPrefix netip.Prefix
+	if ip.Is4() {
+		ipPrefix = netip.PrefixFrom(ip, 32)
+	} else {
+		ipPrefix = netip.PrefixFrom(ip, 128)
+	}
```
```
ipPrefix := netip.PrefixFrom(ip, ip.BitLen())
```

khenidak commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,679 @@
+	// we can't check longer than the request mask
+	mask := prefix.Bits()
+	// walk the network bits of the prefix
+	for bitPosition < mask {
```
Low Pri: There are so many loops/iterators that looks like this one. Why don't we go with a unified implementation a la visitor with function that loops with a function pointer as an arg that performs what is needed?

khenidak commented:
/retest /approve

khenidak commented:
/retest
/lgtm

danwinship commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,566 @@
+	c.serviceCIDRsSynced = serviceCIDRInformer.Informer().HasSynced
+
+	// IPAddresses can only block the deletion of the ServiceCIDR that contains it
+	_, _ = ipAddressInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```
(not clear why that comment is here?)

thockin replied:
+1

danwinship reviewed (approved):
/lgtm
/approve

comments about `IPv4`/`IPv4` vs `CIDR` or `CIDRs` still apply, but we will revisit that later

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,454 @@
+			return condition.Status == metav1.ConditionStatus(metav1.ConditionTrue)
+		}
+	}
+	// if condition is not found assume it is ready to handle scenarios
+	// where the kube-controller-manager is not running
```
> Or is it not possible for the "add Service" and the "delete ServiceCIDR" to be overlapping from the perspective of the apiserver?

actually I described this case in the KEP and didn't implement it https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs#service-ip-reservation, instead of removing the finalizer immediately it delays the removal 

let me fix it tomorrow

aojea replied:
I will add all of this in a follow up

thockin commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -721,3 +721,66 @@ func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList
+			allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, err.Error()))
+		} else {
+			if prefix.Addr() != prefix.Masked().Addr() {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, "wrong CIDR format, IP doesn't match network IP address"))
```
"must be a valid IPv4 CIDR with no host bits specified"

thockin commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -721,3 +721,66 @@ func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, "wrong CIDR format, IP doesn't match network IP address"))
+			}
+			if prefix.String() != cidrConfig.Spec.IPv4 {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, "CIDR not in canonical format"))
```
Do we really need this check?  If it parsed as IPv4, isn't that enough?

thockin commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -721,3 +721,66 @@ func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, "CIDR not in canonical format"))
+			}
+			if !prefix.Addr().Is4() {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, "not IPv4 family CIDR"))
```
"must be a valid IPv4 CIDR"

thockin commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -721,3 +721,66 @@ func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList
+	if cidrConfig.Spec.IPv6 != "" {
+		prefix, err := netip.ParsePrefix(cidrConfig.Spec.IPv6)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("IPv6"), cidrConfig.Spec.IPv6, err.Error()))
```
all same comments

thockin reviewed (commented):
first few commits only

For review sanity it would be best to add commits after the original commits so I can see what changed, before squashing them

thockin commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,219 @@
+			if condition.Status == metav1.ConditionTrue {
+				return
+			}
+			klog.Infof("default ServiceCIDR condition Ready is not True: %v", condition.Status)
```
What does it mean for a CIDR to not be ready?  When would that be set, and what should clients do with that information?

aojea replied:
right now Ready is used to notify users that the Deletion of a ServiceCIDR is blocked because it will leave IP Addresses orphan, and the controller will not remove the finalizer

thockin commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,219 @@
+		}
+	}
+	// set status to ready if the ServiceCIDR matches this configuration
+	if c.ipv4CIDR == serviceCIDR.Spec.IPv4 &&
```
In case of multiple API servers, if just one of them has the wrong config, won't this condition flip back and forth, depending on who wrote it last?

aojea replied:
the condition is only set to true here, in theory (Dan already pointed it out) this controller should not event have to set the condition, but I wanted to be paranoid and cover the case that the controller manager is not running and not setting the condition.

It can not flip because it only goes to true if it matches the conditions, apiservers with different configfs just will exist

aojea commented on pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go:
```diff
@@ -0,0 +1,219 @@
+		svcApply := networkingapiv1alpha1apply.ServiceCIDR(DefaultServiceCIDRName).WithStatus(svcApplyStatus)
+		if _, errApply := c.client.NetworkingV1alpha1().ServiceCIDRs().ApplyStatus(context.Background(), svcApply, metav1.ApplyOptions{FieldManager: controllerName, Force: true}); errApply != nil {
+			klog.Infof("error updating default ServiceCIDR status: %v", errApply)
+			c.eventRecorder.Eventf(serviceCIDR, v1.EventTypeWarning, "KubernetesDefaultServiceCIDRError", "The default ServiceCIDR Status can not be set to Ready=True")
```
I thought that SSA with force handled this https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts , but better ask @apelisse to be completely sure

aojea committed:
remove flag cidr max size validation if gate enable

aojea committed:
update ServiceCIDR API: replace IPv4,IPv6 fields by an array of strings

https: //docs.google.com/document/d/1-CKDPRHsG6XfnGs-Bb7ptWgDtTq9-hj8aJQpGz1CSh8/edit

Change-Id: If7134a5553b4ecc12b1d58ffc52ebb9f8594778a

aojea committed:
fix race on ServiceCIDR deletion

When a ServiceCIDR is deleted, the service CIDR controller on the
controller manager verifies that is safe to be deleted before removing
the finalizer, howerver, since the information of deletion takes time to
propragate, there can be a race where the apiserver allocators didn't
receive the information of deletion and assign an IP address that will
be orphan.

To avoid this race, the service cidr controller waits a grace period
before removing the finalizer to ensure the allocators do not assign any
new IP Address from that range before is completely deleted.

Change-Id: Ib34d32c0bdde91c6e84f1d056db9374589b25c0b

aojea committed:
printers: use new ServiceCIDR API

Change-Id: Iaedeb99fcdb6e27e9987bccb1ccc32fab7da71f6

aojea committed:
describer: use new ServiceCIDR API

Change-Id: Iff11c70f20bab3e55e4e569fb110ef25dd6dd97e

aojea committed:
default_servicecidr_controller: use new ServiceCIDR API

Change-Id: I2c53815136e0d985959420911339aacd62e01e78

aojea committed:
ipallocator/cidrallocator: use new ServiceCIDR API

Change-Id: I5b300373e21cc65f6ef171790ef81ee391f1c752

cidrallocator

Change-Id: I13050a9d60360e555b75a13beaa9a923268ec4e8

aojea committed:
repairip controller: use new ServiceCIDR API

aojea committed:
service cidr controller manager: use new ServiceCIDR API

aojea committed:
test e2e: use new ServiceCIDR API

Change-Id: I82075c4252d8734886f80dd68327fe73e2ed4ba5

aojea committed:
integration etcd data: use new ServiceCIDR API

Change-Id: I68a91815b3220a41e97667d8f8f4a3bdf5a91e8a

aojea committed:
test integration: use new ServiceCIDR API

Change-Id: I7aaa978842fbe8d75383f8a3159ca4fce01ffdcb

aojea committed:
make update with the new API with arrays

thockin commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -726,55 +726,54 @@ var ValidateServiceCIDRName = apimachineryvalidation.NameIsDNSSubdomain
-	if cidrConfig.Spec.IPv4 == "" && cidrConfig.Spec.IPv6 == "" {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("Spec"), cidrConfig.Spec, "at least one CIDR required"))
+	if len(cidrConfig.Spec.CIDRs) == 0 {
+		allErrs = append(allErrs, field.Invalid(fieldPath, cidrConfig.Spec, "at least one CIDR required"))
```
`field.Required()`

thockin commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -726,55 +726,54 @@ var ValidateServiceCIDRName = apimachineryvalidation.NameIsDNSSubdomain
-				allErrs = append(allErrs, field.Invalid(field.NewPath("IPv4"), cidrConfig.Spec.IPv4, "not IPv4 family CIDR"))
-			}
+	if len(cidrConfig.Spec.CIDRs) > 2 {
+		allErrs = append(allErrs, field.Invalid(fieldPath, cidrConfig.Spec, "max two CIDRs"))
```
"may only hold up to 2 values"

Check the error messages for ClusterIPs

thockin commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -726,55 +726,54 @@ var ValidateServiceCIDRName = apimachineryvalidation.NameIsDNSSubdomain
+	if len(cidrConfig.Spec.CIDRs) == 2 {
+		isDual, err := netutils.IsDualStackCIDRStrings(cidrConfig.Spec.CIDRs)
+		if err != nil || !isDual {
+			allErrs = append(allErrs, field.Invalid(fieldPath, cidrConfig.Spec, "must be one CIDR from each IP family, i.e 192.168.0.0/24 and 2001:db8::/64"))
```
"may specify no more than one IP for each IP family"

thockin reviewed (commented):
Just covering the revised API commits

thockin commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,679 @@
+	"net/netip"
+)
+
+// iptree implement a radix tree that uses IP prefixes as nodes and allows to store values in each node.
```
This should be a package comment, shouldn't it?

thockin commented on pkg/util/iptree/iptree.go:
```diff
@@ -0,0 +1,679 @@
+limitations under the License.
+*/
+
+package iptree
```
Do we really need this?  I get that having thousands of CIDRs might be a problem, but I bet that a simple linear search will be at least as good, maybe better, for 99.999% of users, and when someone has enough CIDRs to need this, we can ask them "why?"

aojea replied:
talked offline, in a follow up we'll move to a more naive logic and remove the tree, but since is alpha, the code is feature gated and we are in code freeze , we keep it as is

thockin reviewed (commented):
Just covering the radix tree commit

thockin reviewed (commented):
Comments on allocator, but the main point is that I think this can all be much simpler without the radix tree and a little O(n) code (which is fine because it is small-n).

Discussed in real time.

thockin commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -330,44 +359,40 @@ func (r *RepairIPAddress) syncService(key string) error {
 			// ClusterIP is out of range
-			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]: %s is not within the configured Service CIDR %s; please recreate service", v1Family, ip, network.String())
-			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s is not within the service CIDR %s; please recreate", v1Family, ip, svc.Namespace, svc.Name, network.String()))
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]: %s is not within the configured Service CIDR; please recreate service", family, ip)
```
s/the/any ?

thockin reviewed (commented):
repair loop looks ok

thockin commented on pkg/controller/servicecidrs/servicecidrs_controller.go:
```diff
@@ -0,0 +1,566 @@
+				WithLastTransitionTime(metav1.Now()))
+		svcApply := networkingapiv1alpha1apply.ServiceCIDR(cidr.Name).WithStatus(svcApplyStatus)
+		_, err = c.client.NetworkingV1alpha1().ServiceCIDRs().ApplyStatus(ctx, svcApply, metav1.ApplyOptions{FieldManager: controllerName, Force: true})
+		return err
```
Can you comment here about what will trigger this to be re-evaluated?  Is it a timer or just teh regular periodic full sync?  or something else?

aojea replied:
it is a trigger , the handlers process  ServiceCIDR or IPAddress changes and enqueue the impacted ServiceCIDRs, just to check that change impact on the state of a ServiceCIDR that is going to be deleted

thockin reviewed (commented):
Controller looks OK, but I admit I skimmed the test

aojea commented on pkg/registry/core/service/ipallocator/cidrallocator.go:
```diff
@@ -0,0 +1,454 @@
+
+var _ Interface = &MetaAllocator{}
+
+// NewMetaAllocator returns an IP allocator associated to a network range
```
the allocator is associated to a ServiceCIDR range,

aojea committed:
servicecidrs controller clarify condition false reevaluation

Change-Id: I0eb8d39abe9b7b0ce6472ff426e9a62e7155aae1

aojea commented:
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-conformance-kind-ga-only-parallel

aojea commented:
/test pull-kubernetes-e2e-kind-alpha-features

thockin commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -721,3 +721,65 @@ func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList
+		allErrs = append(allErrs, field.Invalid(fldPath, cidr, err.Error()))
+	} else {
+		if prefix.Addr() != prefix.Masked().Addr() {
+			allErrs = append(allErrs, field.Invalid(fldPath, cidr, "wrong CIDR format, IP doesn't match network IP address"))
```
nit for followup:    I think this can be:  "may not include host bits").

But also, won't this be caught by the string-compare below, which has a more useful error?

thockin commented on pkg/apis/networking/types.go:
```diff
@@ -642,3 +642,53 @@ type IPAddressList struct {
+
+// ServiceCIDRStatus describes the current state of the ServiceCIDR.
+type ServiceCIDRStatus struct {
+	// conditions holds an array of metav1.Condition that describe the state of the ServiceCIDR.
```
I think we should document known Conditions and what they mean (e.g. my question about "ready")

thockin reviewed (commented):
For the sake of code-freeze, this seems sufficiently gated and OK.  Lots to follow up on.

/approve
/lgtm

aojea committed:
hack/update-codegen

Change-Id: I929e8fe0313ec2e0aef2b26a5138cd88bdfe8914

dims commented:
few more things to fix here @aojea

thockin commented:
I'm still LGTM and the tests are still not

aojea commented:
> I'm still LGTM and the tests are still not

the jobs failing are unrelated to this PR, and not required

thockin commented:
/lgtm

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -369,6 +369,16 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
 			eventsRule(),
 		},
 	})
+	addControllerRole(&controllerRoles, &controllerRoleBindings, rbacv1.ClusterRole{
```
this added an unconditional role / binding to default cluster policy

this should be wrapped in a feature gate check so we don't create these permissions until the feature is enabled

fixed in #121682

aojea replied:
thanks for fixing it
